Adding features

kdimitrov03 opened the pull request:
Removed the use of namespace and now we just use the build in import and export. Added two more Project Lists that can store project (To-Do and Review) also removed the ability to drag a project from the Review section and drop it in the To-Do. In addition added some comments to make the code easier to understand.

JustKdd committed:
added to more lists and also comments

kdimitrov03 committed:
Cannot move from review to to do

kdimitrov03 committed:
removed the use of namespace

rkoychev commented on src/app.ts:
```diff
@@ -2,20 +2,15 @@
+//add two more to do and for review from review only to finished or active
+import { ProjectInput } from "./components/project-input.js";
+import { ProjectList } from "./components/project-list.js";
+  //instance of the ProjectInput class which takes care of the user input with validation and event listeners for submitting
```
this comment is more appropriate to be where the ProjectInput class is defined, and not where  it is used. If someone uses it at another place they will not see the comment.

rkoychev commented on src/components/project-list.ts:
```diff
@@ -1,74 +1,97 @@
+  implements DragTarget
+{
+  assignedProjects: Project[] = [];
+  constructor(private type: "active" | "finished" | "to-do" | "review") {
```
We can have a type defined with all 4 options, and use a variable of that type anywhere we reference the project state.

rkoychev commented on src/app.ts:
```diff
@@ -2,20 +2,15 @@
+  //instance of the ProjectInput class which takes care of the user input with validation and event listeners for submitting
   new ProjectInput();
+  //instances of the ProjectList class which stores the project and we initialize them with the type of list
+  new ProjectList("to-do");
```
Extract all the States of a project in a custom type that you can use throughput the project

rkoychev commented on src/components/project-list.ts:
```diff
@@ -1,74 +1,97 @@
+    if (this.type === "active") {
+      status = ProjectStatus.Active;
+    } else if (this.type === "to-do") {
+      if (+prjStatus === ProjectStatus.Review) {
```
what is the plus sign doing here ?

rkoychev commented on src/components/project-list.ts:
```diff
@@ -1,74 +1,97 @@
+      status = ProjectStatus.Active;
+    } else if (this.type === "to-do") {
+      if (+prjStatus === ProjectStatus.Review) {
+        status = ProjectStatus.Review;
```
i see a mix of hardcoded strings and Enums that makes the code a bit confusing

rkoychev commented on src/decorators/autobind.ts:
```diff
@@ -1,18 +1,17 @@
\ No newline at end of file
+//autobind decorator
+//decorator to automatically bind the this keyword method to the object instance
+export function autobind(
```
do you understand what this does? haven't seen such kind of binding. Think about if you can do this in another way. Using arrow functions to register event handlers is usually a much cleaner way to handle this.

rkoychev replied:
If not used remove ( or at least comment ) this code

rkoychev commented on src/models/project.ts:
```diff
@@ -1,15 +1,15 @@
-}
\ No newline at end of file
+export enum ProjectStatus {
+  Active,
```
it is ok to have enum and not a custom type, if you are  leaving the enum i think you need to have string values for each state so you can map those easily in the code

rkoychev commented on src/state/project-state.ts:
```diff
@@ -1,53 +1,53 @@
+type Listener<T> = (items: T[]) => void;
+class State<T> {
+  protected listeners: Listener<T>[] = [];
+  //array of listeners where we store the functions that need to be called whenever the state changes or we add a project
```
great comment, if you can rename the property to something that speaks for itself it will be even greater. So if instead of "listeners" we call it "stateChangeListeners" it will be obvious to most that this is a array of listeners that store functions...  and this will make the comment more or less not needed.

rkoychev commented on src/state/project-state.ts:
```diff
@@ -1,53 +1,53 @@
+export class ProjectState extends State<Project> {
+  private projects: Project[] = [];
+  //singleton we can only have one instance of project state class
+  private static instance: ProjectState;
```
put the comment on top of the class , the comment is about the class ( how it is used) , not the static instance property ( it is hidden from outside the class)

rkoychev commented on src/state/project-state.ts:
```diff
@@ -1,53 +1,53 @@
+    this.instance = new ProjectState();
+    return this.instance;
+  }
+  //when creting a project we add a random id and set it to do
```
Maybe we need a default value that is equal to TODO state and use the default value here

rkoychev commented on src/components/project-input.ts:
```diff
@@ -1,74 +1,77 @@
+    const descriptionValidatable: Validatable = {
+      value: enteredDescription,
+      required: true,
+      minLength: 7,
```
when you showed me the project we had error that we didin't meet the min length requirement, but it didn't say what we didn't meet. IF that is still the case, can you point out in the alert what is the requirement that hasn't been met so the user knows what to do

kdimitrov03 committed:
Made changes

Used webpack for more optimal and faster way for the app to load because it takes all the files we use for importing and creates only one final that needs to be run at the start. Also removed the use of the autobind decorator and i just use the build in .bind() method where needed. In addition removed the enum and its place added a projectStatus type that can have a value of the possible project statuses . Some other minor changes are the change of the name of the Listeners array to stateChangeListeners for easier understanding and made some changes to the comments in the project

kdimitrov03 committed:
Binding change

Removed the bind method and used arrow functions in its place

rkoychev commented on src/components/project-input.ts:
```diff
@@ -1,74 +1,129 @@
+    event.preventDefault();
+    const userInput = this.gatherUserInput();
+    //if user input is correct we receive the array
+    if (Array.isArray(userInput)) {
```
i think we can also do : `if ( userInput instanceOf Array) ` . But your way is also fine

rkoychev commented on src/components/project-input.ts:
```diff
@@ -1,74 +1,129 @@
+    const userInput = this.gatherUserInput();
+    //if user input is correct we receive the array
+    if (Array.isArray(userInput)) {
+      const [title, description, people] = userInput;
```
why are you destructing the userInput just to pass it as parameters on the next line when you can just have one line like 
projectState.addProject(userInput[0], userInput[1], userInput[2]) .
Its not wrong , and it is a bit more descriptive which is fine, but if there is no manipulation of the fields in any way you can just pass them from the array.

rkoychev commented on src/components/project-input.ts:
```diff
@@ -1,74 +1,129 @@
-        projectState.addProject(title, description, people);
-        this.clearUserInputs();
+  }
+  //cleans user input fields
```
the function name is the same as the comment, it is descriptive enough to make the comment redundant

rkoychev commented on src/components/project-input.ts:
```diff
@@ -1,74 +1,129 @@
       }
+      alert(errorText);
+    } else {
+      return [enteredTitle, enteredDescription, Number(enteredPeoples)];
```
if you are trying to cast "enteredPeoples" ( which has a typo btw) this is probably not the best way. Check this article https://stackoverflow.com/questions/14667713/how-to-convert-a-string-to-number-in-typescript

rkoychev replied:
as i can see you are doing it with a "+" on the previous lines. IF you are not sure which is the best way to do something check online or consult a colleague. Doing it in different ways is usually not good

rkoychev reviewed (approved):
You can merge your pr after you address the few issues in the comments

kdimitrov03 committed:
Small changes before merge
